crates/shared/tests/vtag_test.rs

```diff
@@ -258,3 +279,43 @@ fn it_allows_aria_attributes() {
         panic!("vtag expected");
     }
 }
+
+#[test]
+fn it_checks_mixed_closing_tags() {
+    let a: VNode<Comp> = html! {
+        <div>
+            <div/>
+        </div>
+    };
+    let b: VNode<Comp> = html! {
+        <div>
+            <div></div>
+        </div>
+    };
+    assert_eq!(a, b);
+
+    let a: VNode<Comp> = html! {
+        <div>
+            <div data-val={ 2 / 1 }/>
```

Comment:
Can you also add a test for an expression with a `>`? Thanks!

Reply:
This works:
```rust
let a: VNode<Comp> = html! { <div> <div data-val={ 2 > 1 }/>  </div> };
let b: VNode<Comp> = html! { <div> <div data-val={ true }></div> </div> };
assert_eq!(a, b);
```
But this doesn't:
```rust
let a: VNode<CompInt> = html! { <div> <div onblur=|_| 2 > 1/>  </div> };
let b: VNode<CompInt> = html! { <div> <div onblur=|_| false></div> </div> };
assert_eq!(a, b); // NB: assert_eq! doesn't (cannot) compare the closures
```

It wasn't working before this PR, is now breaking with a less nice message. This should however happen very rarely, and the easy workaround is to add brackets around `2 > 1`. React's JSX doesn't have this problem AFAIK because it enforces putting brackets around expressions. I think that I'll take the pragmatic approach here and try to add a meaningful error message.

Reply:
The aforementioned failing test on `master`:

```
diff --git a/tests/vtag_test.rs b/tests/vtag_test.rs
index 5ca5d6d..312888c 100644
--- a/tests/vtag_test.rs
+++ b/tests/vtag_test.rs
@@ -28,6 +28,27 @@ impl Renderable<Comp> for Comp {
     }
 }
 
+struct CompBool;
+
+impl Component for CompBool {
+    type Message = bool;
+    type Properties = ();
+
+    fn create(_: Self::Properties, _: ComponentLink<Self>) -> Self {
+        CompBool
+    }
+
+    fn update(&mut self, _: Self::Message) -> ShouldRender {
+        unimplemented!();
+    }
+}
+
+impl Renderable<CompBool> for CompBool {
+    fn view(&self) -> Html<Self> {
+        unimplemented!();
+    }
+}
+
 #[test]
 fn it_compares_tags() {
     let a: VNode<Comp> = html! {
@@ -258,3 +279,11 @@ fn it_allows_aria_attributes() {
         panic!("vtag expected");
     }
 }
+
+#[test]
+fn this_fails() {
+    let a: VNode<CompBool> = html! { <div> <a onblur=|_| 2 > 1 />  </div> };
+    let b: VNode<CompBool> = html! { <div> <a onblur=|_| false></div> </div> };
+    assert_eq!(a, b);
+}
+

```

```
error: this open tag has no corresponding close tag
   --> tests/vtag_test.rs:285:44
    |
285 |     let a: VNode<CompBool> = html! { <div> <a onblur=|_| 2 > 1 />  </div> };
    |                                            ^^^^^^^^^^^^^^^^^
error: this open tag has no corresponding close tag
   --> tests/vtag_test.rs:286:44
    |
286 |     let b: VNode<CompBool> = html! { <div> <a onblur=|_| false></div> </div> };
    |                                            ^^^^^^^^^^^^^^^^^^^^
error: aborting due to 2 previous errors
error: Could not compile `yew`.
```

Reply:
I had a feeling that was broken in the current macro implementation ;) thanks for adding tests and a better error message!

Reply:
I see three choices in order to provide better error messages:
* wait for [`proc-macro` diagnostics](https://doc.rust-lang.org/proc_macro/struct.Diagnostic.html), which are currently nightly-only, and just use `eprintln!` in the meantime (this error should occur quite rarely).
* mess around the actual code in order to return a result from verify_end.
* use global variables.

I implemented the first one for now. What do you think?